parsely/src/main/java/com/parsely/parselyandroid/AdvertisementIdProvider.kt

```diff
@@ -0,0 +1,47 @@
+package com.parsely.parselyandroid
+
+import android.content.Context
+import android.provider.Settings
+import com.google.android.gms.ads.identifier.AdvertisingIdClient
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.launch
+
+internal class AdvertisementIdProvider(
```

Comment:
I am curious about the general design choice for this class.

If I am understanding it correctly, when an object of this class is instantiated, if we were to read the `adKey` immediately, it'll return `null`. Looking at the constructor signature, there is a hint that we are using coroutines for _something_ but there isn't much context about what it's used for unless we look into the implementation. So, I think a developer may not realize that gotcha and end up requesting the `adKey` before it's ready. Also, I am not even sure if there is a straightforward way to tell when the `adKey` is ready to be retrieved.

So, I was wondering if we should use a `suspend` function for `provide` so that it can be used in a proper coroutine scope without worrying about the internal details. What do you think?

Reply:
I think this one is important to get right. We don't have to use a suspended function if it's going to be more trouble than it's worth - but at the very least, I think it's worth having a big disclaimer in both the class definition and where it's consumed. The reason I strongly feel this way is because if we have any issues related to this, it can be hard to debug if the developer doesn't have any prior context. To be fair - the fact that it's marked as `nullable` may give enough context, so maybe it's not as big of a deal as I am making it out to be 🤷‍♂️ 

I don't think we should make any major changes in this PR, because it'll get hard to review it. However, if I was working on this, I'd give using suspended function a chance as a separate PR. Assuming we keep the current architecture, this will cause a chain reaction and the following methods may also need to become suspended functions: `AndroidDeviceInfoRepository.collectDeviceInfo`, `EventsBuilder.buildEvent` & `ParselyTracker.trackPlay`.

* `ParselyTracker.trackPlay`: Looking at the name, I assumed that this function would eventually become a suspended function regardless of the decision we'll make here, but looking at the implementation, I am not so sure. I'd love to hear your thoughts!
* `EventsBuilder.buildEvent`: This probably shouldn't be a suspended function, but instead take the device info as a parameter. This is used in `ParselyTracker.trackPlay` & `ParselyTracker.trackPageview` both of which _could_ be suspended functions, so I think that'll work out.
* `AndroidDeviceInfoRepository.collectDeviceInfo`: Making this a suspended function makes sense to me. It communicates that collecting device info is not an immediate thing which could be for many reasons. We could be reading from file, checking a DB etc to collect this info. Admittedly some of the info is available immediately, so I could see an argument for keeping it a regular function as well. 🤷‍♂️ 

I think the decision will heavily depend on your vision for the SDK in general and I don't want to influence that vision too much. If you think it's better to leave the current design as is, I can live with that as long as we document it.

Hope you find my thoughts useful - let me know what you think!

Reply:
> As that's the case - would you mind accepting this PR, or do you think we should iterate on something more here?

I just wanted this conversation to resolve before approving the PR - in case you wanted to add a comment about this issue. Approved now - thanks for the discussion!